packages/massa-web3/src/experimental/basicElements/operationManager.ts

```diff
@@ -0,0 +1,286 @@
+          unsigned.encode(BigInt(operation.contractDataBinary.length))
+        )
+        components.push(operation.contractDataBinary)
+
+        operation.datastore =
+          operation.datastore || new Map<Uint8Array, Uint8Array>()
+        components.push(unsigned.encode(BigInt(operation.datastore.size)))
+
+        // length prefixed key-value pairs encoding
+        for (const [key, value] of operation.datastore) {
+          const keyBytes = Buffer.from(key)
+          const keyLen = unsigned.encode(BigInt(keyBytes.length))
+          const valueBytes = Buffer.from(value)
+          const valueLen = unsigned.encode(BigInt(valueBytes.length))
+          components.push(keyLen, keyBytes, valueLen, valueBytes)
+        }
+        break
+      default:
+        throw new Error('Operation type not supported')
+    }
+
+    return Uint8Array.from(
+      components.flatMap((component) => Array.from(component))
+    )
+  }
+
+  /**
+   * Deserializes an operation according to the Massa protocol.
+   *
+   * @param data - The byte array to deserialize.
+   *
+   * @returns An new instance of OperationDetails representing the deserialized operation.
+   */
+  static deserialize(data: Uint8Array): OperationDetails {
+    let offset = 0
+    const decodeNext = () => {
+      const value = unsigned.decode(data, offset)
+      offset += unsigned.encodingLength(value)
+      return Number(value)
+    }
+
+    let operationDetails: BaseOperation = {
+      fee: decodeNext(),
+      expirePeriod: decodeNext(),
+      type: decodeNext(),
+    }
+
+    switch (operationDetails.type) {
+      case OperationType.Transaction: {
+        const recipientAddress = Address.fromVersionedBytes(
+          data.slice(offset, offset + 33)
+        )
+        offset += 33
+        const amount = decodeNext()
+        return {
+          ...operationDetails,
+          recipientAddress,
+          amount,
+        } as TransferOperation
+
+        break
+      }
+      case OperationType.RollBuy:
+      case OperationType.RollSell: {
+        const amount = decodeNext()
+        return {
+          ...operationDetails,
+          amount,
+        } as RollOperation
+        break
+      }
+      default:
+        throw new Error('Operation type not supported')
+    }
+  }
+
+  /**
+   * Formats an operation for signing.
+   *
+   * @param chainId - The identifier of the blockchain network.
+   * @param operation - The operation to sign.
+   * @param key - The public key to sign the operation with.
+   *
+   * @returns The formatted operation ready for signing.
+   */
+  static canonicalize(
+    chainId: number,
+    operation: OperationDetails,
+    key: PublicKey
+  ): Uint8Array {
+    // u64ToBytes is little endian
+    const networkId = new Uint8Array(8)
+    const view = new DataView(networkId.buffer)
+    view.setBigUint64(0, BigInt(chainId), false)
+
+    const data = OperationManager.serialize(operation)
+    const publicKeyBytes = key.versionedBytes()
```

Comment:
if we don't have an other `toBytes()` than `versionedBytes` maybe we should rename it to `toBytes()` and use this only representation as byte everywhere

Reply:
My point is that there is no other toBytes in this package

Reply:
Sorry for the misunderstanding. When dealing with `toBytes` functions on cryptographic objects, I'm expecting to have something that represents a *real* key, which is not the case here. Having fallen into this pitfall during some implementations, I went with a clearer terminology to help the next developer who works on this.

Furthermore, the end-user developer should never deal with this function directly.

This is why I've decided to offer a more verbose, but clearer API.